common/edl/asym_keys.edl

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+enclave
+{
+    /**
+     * This enumeration defines the type of a asymmetric key.
+     * This definition is shared by the enclave and the host.
+     */
+    enum oe_asymmetric_key_type_t {
+        /**
+         * A secp256r1/NIST P-256 elliptic curve key.
+         */
+        OE_ASYMMETRIC_KEY_EC_SECP256P1 = 1,
+
+        /**
+         * Unused.
+         */
+        _OE_ASYMMETRIC_KEY_TYPE_MAX = 4294967295
+    };
+
+    /**
+     * This enumeration defines the format of the asymmetric key.
+     * This definition is shared by the enclave and the host.
+     */
+    enum oe_asymmetric_key_format_t {
+        /**
+         * The PEM format.
+         */
+        OE_ASYMMETRIC_KEY_PEM = 1,
+
+        /**
+         * Unused.
+         */
+        _OE_ASYMMETRIC_KEY_FORMAT_MAX = 4294967295
+    };
+
+    /**
+     * This struct contains the parameters for asymmetric key derivation.
+     * This definition is shared by the enclave and the host.
+     */
+    struct oe_asymmetric_key_params_t {
+        /**
+         *  The type of asymmetric key.
+         */
+        oe_asymmetric_key_type_t type;
+
+        /**
+         * The exported format of the key.
+         */
+        oe_asymmetric_key_format_t format;
+
+        /**
+         * Optional user data to add to the key derivation.
+         */
+        [size=user_data_size] void* user_data;
```

Comment:
Given that `user_data` is marshaled by oeedger8r, I suspect there will be an accompanying change to the code that was consuming `oe_assymmetric_key_params_t`. It would no longer need to do boundary validation on `user_data`.

Reply:
The annotation here is specifically for deep copy.
For example, https://github.com/openenclave/openenclave/blob/74c7a6f9f93bff670c2e6d5ec0548552da5012c8/tests/oeedger8r/edl/deepcopy.edl#L10 will result in shallow copy, while https://github.com/openenclave/openenclave/blob/74c7a6f9f93bff670c2e6d5ec0548552da5012c8/tests/oeedger8r/edl/deepcopy.edl#L22 will trigger the deep copy based on the `count`.

Reply:
What I meant was that previously, the `user_data` field which is a nested pointer was being shallow-copied (equivalent t o `user_check` annotation). With this PR, it is being deep-copied.
Since `oe_assymmetric_key_params_t` was being passed as input to ecalls, then
- Either we were doing a check `oe_is_outside_enclave(params->user_data, params->user_data_size)`. If so this check can be removed now since edger8r does a deep copy.
- Or we weren't doing the above check, then it means that there was a potential security vulnarability since the host could set `params->user_data` to point to a location within the enclave. In that case, this PR fixes that vulnerability and it would be good to call that out in the PR.

Reply:
@anakrish This enclave-side code should have been checking that the memory is all in the host, but it wasn't. So no corresponding code change is needed here.

But this is all the more reason to not use foreign structures.

Reply:
Interestingly, we didn't have those checks. I'll look more into that and decide whether to address in this PR or to open another issue for that.

Reply:
@jhand2 Thanks for confirming that this is fixing a security vulnerability. It is important to be call that out in the PR comments. Interesting that the edge-function dealing with cryptography was not fully secure.

Yes, this strongly underlines the need to use fully marshalled EDL. I think in some places we use explicit `user_check` annotation. Hopefully once the clean up is done, there will just be a handful of `user_check` and everything else is fully marshaled.

Reply:
@swamymsft This PR fixes a security vulnerability. As @jhand2 mentioned, EDL specifications ought not to use foreign structures. Where would the vulnerability being fixed land in your bug bar #2634?

Reply:
@anakrish Reading the discussion it looks like the arbitrary enclave memory can be copied to host by specifying any address and size, is that correct?

Reply:
Please see  the analysis as follows. The conclusion is that this seems no security vulnerability. However, we can perform the check much earlier.
 
`oe_asymmetric_key_params_t` is used by two ecalls: `oe_get_public_key_ecall` and `oe_get_public_key_by_policy_ecall`.
The current behavior of argument passing is shallow copy (with deepcopy turned off), meaning that the host pointer will be passed into the enclave.

The `user_data` and `user_data_size` is used as the following trace.
`oe_get_public_key_ecall`
-> `oe_get_public_key`
-> `_load_asymmetric_key`
-> `_derive_asymmetric_key`
-> `_create_asymmetric_keypair`
-> `oe_kdf_derive_key` (`user_data` and `user_data_size` are passed as arguments)
-> `kdf_hmac_sha256_ctr`
-> `oe_hmac_sha256_update`
At this point, the size will be checked against `OE_INT_MAX`:
```
if (!context || !data || size > OE_INT_MAX)
        OE_RAISE(OE_INVALID_PARAMETER);
```

Reply:
@swamymsft It is write to an arbitrary location in the enclave, but not arbitrary data. The host can make the enclave write a public key to any enclave address, but the host doesn't have full control over the contents of that key (only partial control).

Reply:
thanks @jhand2 & @mingweishih, this isn't as worse as I initially thought. But I agree with @jhand2 that it can be used to infer the contents of the enclave memory depending on the key derivation.

If the key derivation deterministically yield to same value for a particular value in the memory, then an attacker can use it to actually know the arbitrary enclave memory value by repeatedly feeding different values through a host memory pointer that he control. In this case the bug can be converted to arbitrary enclave memory read. The severity depends on the key derivation function, any insights into this function?